### DISCUSSION TITLE
Chaining callbacks

### DISCUSSION THREAD

**Discussion by battjt** [fltk-rs/fltk-rs#865] [Q&A]
I'm new to rust and to fltk, but I think the following should work.  Coming from Swing, I'm used to being able to add multiple Listeners, so I thought that chaining callbacks would be similar.  I based this work on the example in https://docs.rs/fltk/1.1.13/fltk/prelude/trait.WidgetExt.html#tymethod.callback
```
pub trait Callbacks {
    fn add_callback<F: FnMut(&mut Self) + 'static>(&mut self, cb: F)
    where
        Self: Sized;
}
impl<T> Callbacks for T
where
    T: WidgetExt,
    T: Sized,
{
    fn add_callback<F: FnMut(&mut Self) + 'static>(&mut self, mut callback: F)
    where
        Self: Sized,
    {
        let mut cb = self.callback();
        self.set_callback(move |widget| {
            callback(widget);
            if let Some(func) = cb.as_mut() {
                (*func)();
            }
        });
    }
}
```
But it always seg faults on ```(*func)();```  Also I notice that self.callback() is always Some.  Shouldn't it be None if I haven't set a callback yet?

Here is an example project: https://github.com/battjt/cb-test

**Comment by MoAlyousef** [comment]
Hi
Unfortunately method chaining isn't supported, you'd have to create a widget wrapper and include some sort of queue. I'm surprised Rust isn't detecting that these are None values, or if FLTK is just not passing nulls for the callback method.

But yeah I'll add a check to see if a widget has user_data (i.e a callback defined on the Rust side, in that case it'll return None). 

Due to the nature of passing closures from Rust to C/C++, the closed data (here the whole closure passed in set_callback()) is passed in the void pointer argument, and for which you would have to call the user_data() method to get back, but even then, every set_callback() call deletes the previous user_data() as to avoid memory leaks, so that also won't work for chaining callbacks. 

the callback() function works for FLTK types which already have a defined callback like the scrollbar, it wouldn't work for Rust fltk types which were set using set_callback.

**Reply by MoAlyousef** [reply]
An example of chaining:
```rust
use std::cell::RefCell;
use std::rc::Rc;

use fltk::{app::App, button::Button, prelude::*, window::Window, *};

type ButtonCb = Box<dyn FnMut(&mut Button)>;

pub struct MyButton {
    btn: Button,
    v: Rc<RefCell<Vec<ButtonCb>>>,
}

impl MyButton {
    pub fn new(x: i32, y: i32, w: i32, h: i32, label: &str) -> Self {
        let mut btn = Button::new(x, y, w, h, None).with_label(label);
        let v: Rc<RefCell<Vec<ButtonCb>>> = Rc::from(RefCell::from(Vec::new()));
        btn.set_callback({
            let v = v.clone();
            move |b| {
                for f in v.borrow_mut().iter_mut() {
                    f(b);
                }
            }
        });
        Self { btn, v }
    }
    pub fn add_callback(&mut self, f: impl FnMut(&mut Button) + 'static) {
        self.v.borrow_mut().push(Box::new(f));
    }
}

widget_extends!(MyButton, Button, btn);

pub fn main() {
    let application = App::default();
    let mut window = Window::default();
    window.set_size(110, 110);
    let mut b = MyButton::new(5, 5, 100, 100, "click me");
    let click_rc = Rc::from(RefCell::from(0));
    let click = click_rc.clone();
    b.add_callback(move |b| {
        b.set_label(&format!("click {}", *click.borrow()));
        *click.borrow_mut() += 1;
    });
    b.add_callback(move |_b| {
        println!("click {}", click_rc.borrow());
    });
    window.end();
    window.show();
    application.run().unwrap();
}

```

**Reply by battjt** [reply]
Should set_callback return the old callback, more like a swap?  Again I'm new ot Rust, but wouldn't that let Rust handle the free then?

**Reply by MoAlyousef** [reply]
It was that way for some time early on, it was complicated since the crate either provided a Drop impl for the struct wrapping the callback, in which case it would sometimes delete data which was stack allocated (UB), or you would ask users to delete their data manually which was also error prone (leaks or UB).
